### DISCUSSION TITLE
Support multiple .forceignore files in a project

### DISCUSSION THREAD

**Discussion by shetzel** [forcedotcom/cli#2777] [Ideas]
This discussion is based on the issue raised here: [.forceigore in submodules overrides parent](https://github.com/forcedotcom/cli/issues/2758)

Question 1: How desirable is this feature?

Question 2: How should this be supported?  The assumption is, "do what git does", but let us know if you have other ideas.  Keep in mind that this behavior happens within the `@salesforce/source-deploy-retrieve` library that doesn't require a project.

Question 3: What are your usecases?

**Comment by k-capehart** [comment]
Having it behave the same way as .gitignore files seems like the best option to me. It's intuitive and I can't think of any reason why it should be different. Although I'm interested in what the complications are that you mentioned for the `@salesforce/source-deploy-retrieve` library.

**Reply by shetzel** [reply]
The complications are:

- The SDR library does not need to operate within a DX project whereas git always operates within a git repo and can assume things based on that.  E.g., a project root directory cannot be assumed.
- The current code in SDR applies a .forceignore for each metadata component and that is where searching up begins.  That behavior has not changed for many years and might be difficult to modify.
- Making this change could break customers relying on the current behavior, so it would need to be done in a backward compatible way.

**Reply by k-capehart** [reply]
@shetzel I just stumbled across this post again because I have another use case.

We have a CI/CD pipeline where it would be helpful to be able to designate different `.forceignore` files for prod and sandbox deployments.
- `sf project deploy start` - default behavior uses the current traversal method of identifying the `.forceignore` file
- `sf project deploy start --ignore-file=.sandboxforceignore` - new flag to specify a specific file to use for `.forceignore`

This is useful to us because we could ignore certain metadata types when deploying to prod that we don't want to overwrite, like picklist values that might be modified by admins, while still being able to propagate picklist values into sandboxes when changes are made in source control.

**Comment by azlam-abdulsalam** [comment]
@shetzel  There are situations when we want forceignores have to different behavior while building a package, vs deploying to developer environment. This is one approach we do in a flxbl project

https://docs.flxbl.io/sfp/building-artifacts/controlling-aspects-of-the-build-command/selective-ignoring-of-components-from-being-built

While this approach works relatively well, the con being the duplication involved, a better schematic would be kind of override behavior (merge forceignores) and a command or api to display what is getting ignored

**Comment by nwcm** [comment]
The current behaviour of .forceignore is very confusing. In fact, it looks like it will always use the .forceignore from the root of the project everytime. Regardless of if there is a .forceignore in a subdirectory 

take 
- src/package/.forceignore
- src/.forceignore
- .forceignore



## .forceignore
```
*.jsconfig
```

## src/package/.forceignore
```
**__tests__**
```

## result
In this case there are files in __tests__ folders
```
sf project list ignored --source-dir ./src/package
Found the following ignored files:
.forceignore
```
